explorer/interpreter/action.cpp

```diff
@@ -44,32 +46,38 @@ void RuntimeScope::Print(llvm::raw_ostream& out) const {
   out << "}";
 }
 
-void RuntimeScope::Bind(ValueNodeView value_node, Nonnull<const Value*> value) {
+void RuntimeScope::Bind(ValueNodeView value_node, Address address) {
+  CARBON_CHECK(!value_node.constant_value().has_value());
+  auto [it, success] =
+      locals_.insert({value_node, heap_->arena().New<LocationValue>(address)});
+  CARBON_CHECK(success) << "Duplicate definition of " << value_node.base();
+}
+
+void RuntimeScope::BindLifetimeToScope(Address address) {
+  CARBON_CHECK(address.element_path_.IsEmpty())
+      << "Cannot extend lifetime of a specific sub-element";
+  allocations_.push_back(address.allocation_);
+}
+
+void RuntimeScope::BindValue(ValueNodeView value_node,
+                             Nonnull<const Value*> value) {
   CARBON_CHECK(!value_node.constant_value().has_value());
   CARBON_CHECK(value->kind() != Value::Kind::LocationValue);
-  auto allocation_id = heap_->GetAllocationId(value);
-  if (!allocation_id) {
-    auto id = heap_->AllocateValue(value);
-    auto [it, success] = locals_.insert(
-        {value_node, heap_->arena().New<LocationValue>(Address(id))});
-    CARBON_CHECK(success) << "Duplicate definition of " << value_node.base();
-  } else {
-    auto [it, success] = locals_.insert(
-        {value_node,
-         heap_->arena().New<LocationValue>(Address(*allocation_id))});
-    CARBON_CHECK(success) << "Duplicate definition of " << value_node.base();
-  }
+  auto [it, success] = locals_.insert({value_node, value});
```

Comment:
```suggestion
  bool success = locals_.insert({value_node, value}).second;
```

And I see you're just moving this around, but given you're modifying the lines I'm still going to suggest it. Otherwise I look for how `it` is used.

Reply:
Done